### ISSUE TITLE
[branch:dev-refactor] Call to `BaseDataSlice::Init`

### ISSUE THREAD

**Issue by bkj** [gunrock/gunrock#370]
Calling `BaseDataSlice::Init` like
```
GUARD_CU(BaseDataSlice::Init(sub_graph, gpu_idx, target, flag));
```
as in the template [here](https://github.com/gunrock/gunrock/blob/dev-refactor/gunrock/app/Template/Template_problem.cuh#L134) appears to get parsed incorrectly [here](https://github.com/gunrock/gunrock/blob/dev-refactor/gunrock/app/problem_base.cuh#L98): `gpu_idx` gets assigned to `num_gpus`, etc.  

I think the line in the template is supposed to look like [this](https://github.com/gunrock/gunrock/blob/dev-refactor/gunrock/app/pr/pr_problem.cuh#L229), where `num_gpus` is passed into `DataSlice::Init` and then into `DataSliceBase::Init`.

Removing the default arguments in `DataSliceBase::Init` would help avoid this problem by throwing a compiler error if the number of arguments is incorrect.

~ Ben

**Comment by jowens** [comment]
@sgpyc can you comment on this please?

**Comment by neoblizz** [comment]
You are right, I can issue a fix for this.

> Removing the default arguments in DataSliceBase::Init would help avoid this problem by throwing a compiler error if the number of arguments is incorrect.

I have had some issues with this as well, having default parameters could cause confusion specially when a lot of the stuff is templated. I think on the other hand, it is also a common practice and handy sometimes to expect from user to not type in literally every parameter if it is a simple NULL, 0 or a default value. I am iffy about this.

**Comment by sgpyc** [comment]
Parameter matching is terrible, and I spent quite sometime doing that, but couldn't catch all the mismatches.
I really hope C/C++ can have named parameters, but...
A workaround is using a structure to hold some parameters (like what oprtr_parameters does). But it will be too much if using parameter structures for a large number of functions.

If anything works better, I would love to know and try

**Comment by bkj** [comment]
Should I open a quick PR for this?

**Comment by neoblizz** [comment]
@bkj yes please!

**Comment by neoblizz** [comment]
Solved with PR #381, thanks @bkj!
